integrations/gitlab_v2/main.py

```diff
@@ -0,0 +1,157 @@
+
+    if not ocean_action:
+        logger.info(f"Webhook event '{webhook_event}' not recognized.")
+        return {"ok": True}
+
+    if webhook_event == WebHookEventType.MERGE_REQUEST:
+        payload = extract_merge_request_payload(data)
+        await ocean_action(ResourceKind.MERGE_REQUEST, [payload])
+    elif webhook_event == WebHookEventType.ISSUE:
+        payload = extract_issue_payload(data)
+        logger.info(f"Upserting issue with payload: {payload}")
+        await ocean_action(ResourceKind.ISSUE, [payload])
+    else:
+        logger.info(f"Unhandled webhook event type: {webhook_event}")
+        return {"ok": True}
+
+    logger.info(f"Webhook event '{webhook_event}' processed successfully.")
+    return {"ok": True}
+
+
+@ocean.router.post("/webhook")
+async def handle_webhook_request(data: dict[str, Any]) -> dict[str, Any]:
+    webhook_event = data.get("event_type", "")
+    object_attributes_action = data.get("object_attributes", {}).get("action", "")
+    logger.info(
+        f"Received webhook event: {webhook_event} with action: {object_attributes_action}"
+    )
+
+    return await handle_webhook_event(webhook_event, object_attributes_action, data)
+
+
+@ocean.on_resync(ResourceKind.PROJECT)
+async def resync_project(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    tokens = ocean.integration_config["gitlab_access_tokens"]["tokens"]
+    for token in tokens:
+        client = initialize_client(token)
+        config = typing.cast(GitlabProjectResourceConfig, event.resource_config)
+
+        async for projects in client.get_projects():
+            logger.info(f"Received {kind} batch with {len(projects)} projects")
+            if config.selector.onlyGrouped:
+                projects = [project for project in projects if project.get("__group")]
+            yield projects
+
+
+@ocean.on_resync(ResourceKind.GROUP)
+async def resync_group(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    tokens = ocean.integration_config["gitlab_access_tokens"]["tokens"]
+    for token in tokens:
+        client = initialize_client(token)
+        async for groups in client.get_groups():
+            logger.info(f"Received {kind} batch with {len(groups)} groups")
+            yield groups
+
+
+@ocean.on_resync(ResourceKind.MERGE_REQUEST)
+async def resync_merge_request(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    tokens = ocean.integration_config["gitlab_access_tokens"]["tokens"]
+    for token in tokens:
+        client = initialize_client(token)
+        async for merge_requests in client.get_merge_requests():
+            logger.info(f"Received {kind} batch with {len(merge_requests)} merge requests")
+            yield merge_requests
+
+
+@ocean.on_resync(ResourceKind.ISSUE)
+async def resync_issue(kind: str) -> ASYNC_GENERATOR_RESYNC_TYPE:
+    tokens = ocean.integration_config["gitlab_access_tokens"]["tokens"]
+    for token in tokens:
+        client = initialize_client(token)
+        async for issues in client.get_issues():
```

Comment:
Remove the token loops, have a function that inits client and the consider combining all the resync functions into a generic one. It should maybe directly pass the kind to the right method in client that will take the name of the kind and build the uri with it. See octopus integration [here](https://github.com/port-labs/ocean/blob/2f09b26faee005dd91c92e97e4d769e25d6b1878/integrations/octopus/main.py#L64-L70) for ideas